# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Marking.cpp
# Commit: 43978ebfc7ac
# Full Hash: 43978ebfc7ace2c37aba0b7644ac4e866e0730bd
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2021-01-20 16:13:57
# Regressor Bug: 1686219
# File Overlap Count: 1
# Description:
#   Bug 1686219 - Store chunk information at the start rather than the end of the chunk r=sfink
#   
#   This moves the chunk metadata to the start of the chunk and defines the data structures in the public header.  This simplifies accessing this data and removes the need for hardcoded offsets.
#   
#   Requesting review from jandem for JIT updates.
# ==============================================================================

diff -r 496d47ff3fa1 -r 43978ebfc7ac js/src/gc/Marking.cpp
--- a/js/src/gc/Marking.cpp	Wed Jan 20 10:04:11 2021 +0000
+++ b/js/src/gc/Marking.cpp	Wed Jan 20 10:04:29 2021 +0000
@@ -3143,8 +3143,11 @@
 }
 
 #ifdef DEBUG
+static inline uintptr_t OffsetFromChunkStart(void* p) {
+  return uintptr_t(p) & gc::ChunkMask;
+}
 static inline ptrdiff_t OffsetToChunkEnd(void* p) {
-  return ChunkLocationOffset - (uintptr_t(p) & gc::ChunkMask);
+  return ChunkSize - (uintptr_t(p) & gc::ChunkMask);
 }
 #endif
 
@@ -3213,6 +3216,7 @@
   tenuredCells++;
 
   // Copy the Cell contents.
+  MOZ_ASSERT(OffsetFromChunkStart(src) >= sizeof(ChunkHeader));
   MOZ_ASSERT(OffsetToChunkEnd(src) >= ptrdiff_t(srcSize));
   js_memcpy(dst, src, srcSize);
 
@@ -3260,6 +3264,7 @@
   tenuredCells++;
 
   // Copy the Cell contents.
+  MOZ_ASSERT(OffsetFromChunkStart(src) >= sizeof(ChunkHeader));
   MOZ_ASSERT(OffsetToChunkEnd(src) >= ptrdiff_t(srcSize));
   js_memcpy(dst, src, srcSize);
 
@@ -4101,8 +4106,8 @@
 
   MarkBitmapWord* wordp;
   uintptr_t mask;
-  js::gc::detail::GetGCThingMarkWordAndMask(uintptr_t(cell), ColorBit::BlackBit,
-                                            &wordp, &mask);
+  ChunkBitmap::GetMarkWordAndMask(&cell->asTenured(), ColorBit::BlackBit,
+                                  &wordp, &mask);
   return reinterpret_cast<uintptr_t*>(wordp);
 }
 
@@ -4116,8 +4121,7 @@
   ColorBit bit = colorBit == 0 ? ColorBit::BlackBit : ColorBit::GrayOrBlackBit;
   MarkBitmapWord* wordp;
   uintptr_t mask;
-  js::gc::detail::GetGCThingMarkWordAndMask(uintptr_t(cell), bit, &wordp,
-                                            &mask);
+  ChunkBitmap::GetMarkWordAndMask(&cell->asTenured(), bit, &wordp, &mask);
   return mask;
 }
 